From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: callhome:  fix broken proc interface and activate compid
References: bnc#594473
Patch-mainline: Upstream submission made

Symptom:     The proc interface of the driver did not accept valid input
Problem:     strict_strtoul() requires _either_'\n\0' _or_ '\0' termination.
             This was missing and therefore the interface did not recognise
             valid input. The second change ensures that a SLES on z system in
             case of a kernel panic can report the crash to the corresponding
             PMR queue within the IBM service organization (if enabled and
             configured properly).
Solution:    Terminate String and add compid.

Acked-by: John Jolly <jjolly@suse.de>

---
 drivers/s390/char/sclp_async.c |    7 ++-----
 1 file changed, 2 insertions(+), 5 deletions(-)

--- a/drivers/s390/char/sclp_async.c
+++ b/drivers/s390/char/sclp_async.c
@@ -84,6 +84,7 @@ static int proc_handler_callhome(struct 
 		rc = copy_from_user(buf, buffer, sizeof(buf));
 		if (rc != 0)
 			return -EFAULT;
+		buf[sizeof(buf) - 1] = '\0';
 		if (strict_strtoul(buf, 0, &val) != 0)
 			return -EINVAL;
 		if (val != 0 && val != 1)
@@ -135,11 +136,7 @@ static int sclp_async_send_wait(char *me
 	request->sccb = sccb;
 	request->status = SCLP_REQ_FILLED;
 	strncpy(sccb->evbuf.data, message, sizeof(sccb->evbuf.data));
-	/*
-	 * Retain Queue
-	 * e.g. 5639CC140 500 Red Hat RHEL5 Linux for zSeries (RHEL AS)
-	 */
-	strncpy(sccb->evbuf.comp_id, "000000000", sizeof(sccb->evbuf.comp_id));
+	strncpy(sccb->evbuf.comp_id, "5639H7CH0", sizeof(sccb->evbuf.comp_id));
 	sccb->evbuf.header.length = sizeof(sccb->evbuf);
 	sccb->header.length = sizeof(sccb->evbuf) + sizeof(sccb->header);
 	sccb->header.function_code = SCLP_NORMAL_WRITE;
